feat(snql): SnQL Grammar and Visitor

ayirr7 opened the pull request:
This is a second draft attempt at SnQL Grammar, Visitor 

- Much of the visitor and grammar is similar to the existing Clickhouse equivalents
- This currently parses COLLECT (and potentially also) ORDER BY clauses, producing a valid Query object at the root

Next:
- Working on the grammar and visitor functions for other types of clauses
- Adding unit tests

ayirr7 committed:
Draft SnQL grammar with visitor. Handles COLLECT, ORDER BY clauses.

ayirr7 committed:
Merge branch 'master' into feat/snqlvisitor

fpacifici commented on snuba/query/snql/__init__.py:
```diff
@@ -1,23 +1,280 @@
+    collect_clause        = space* "COLLECT" collect_list space*
+    order_by_clause       = space* "ORDER BY" collect_list ("ASC"/"DESC") space*
+
+    collect_list          = collect_columns* (low_pri_arithmetic)
```
Maybe you can just move the content of this one directly into the `collect_clause` above? It does not seem it needed to be split out in its own rule.

fpacifici replied:
I am still not sure I get why you need to split `collect_list` from `collect_columns`. Can't `collect_list` just be `(low_pro_arithmetic space* comma space*)* low_pri_arithmetic` ?

fpacifici commented on snuba/query/snql/__init__.py:
```diff
@@ -1,23 +1,280 @@
-    collect_clause = space* "COLLECT" clause "BY" clause space*
-    having_clause = space* "HAVING" clause space*
-    order_by_clause = space* "ORDER BY" clause ("ASC"/"DESC") space*
+    low_pri_arithmetic    = space* high_pri_arithmetic space* (low_pri_tuple)*
+    high_pri_arithmetic   = space* arithmetic_term space* (high_pri_tuple)*
+    low_pri_tuple         = low_pri_op high_pri_arithmetic
+    high_pri_tuple        = high_pri_op arithmetic_term
+    arithmetic_term       = (space*) (function_call / numeric_literal / column_name) (space*)
+    low_pri_op            = "+" / "-"
+    high_pri_op           = "/" / "*"
+    param_expression      = low_pri_arithmetic / quoted_literal
+    parameters_list       = parameter* (param_expression)
+    parameter             = param_expression space* comma space*
+    function_call         = function_name open_paren parameters_list? close_paren (open_paren parameters_list? close_paren)?
+    simple_term           = quoted_literal / numeric_literal / column_name
+    literal               = ~r"[a-zA-Z0-9_\.:-]+"
+    quoted_literal        = "'" string_literal "'"
+    string_literal        = ~r"[a-zA-Z0-9_\.\+\*\/:-]*"
+    numeric_literal       = ~r"-?[0-9]+(\.[0-9]+)?(e[\+\-][0-9]+)?"
+    column_name           = ~r"[a-zA-Z_][a-zA-Z0-9_\.]*"
+    function_name         = ~r"[a-zA-Z_][a-zA-Z0-9_]*"
+    open_paren            = "("
+    close_paren           = ")"
+    space                 = " "
```
This part is the same we had in the clickhouse parser in the old language. Is that correct or did you make changes?

fpacifici commented on snuba/query/snql/__init__.py:
```diff
@@ -1,23 +1,280 @@
+        columns, _, _, _, = visited_children
+        return columns
+
+    def visit_collect_list(self, node, visited_children):
```
type hints please

fpacifici commented on snuba/query/snql/__init__.py:
```diff
@@ -1,23 +1,280 @@
+    return term
+
+
+class SnQLVisitor(NodeVisitor):
```
You are right, there is a lot of duplication between here and the old expression visitor, I would advise to move that to its own class:
1) create an ExpressionVisitor (do not inherit NodeVisitor), it just contains the methods that produce the expressions
2) move all the common visit_* methods and the classes aruond them into this class.
3) reuse such class from both the visit_methods of this one and the old clickhouse visitor by delegating calls from the visit methods in this class to this external visitor.

I would advise against making this visitor depend directly on the old one. You would have to manage the instance of such visitor, which is something you cannot do reliable (it extends NodeVisitor and the state is handled by Parsimonious).

fpacifici reviewed (changes requested):
This is great, thanks. 
Yeah I agree we can remove a lot of duplication, it should not be hard. See my comment inline about it

ayirr7 committed:
Added ExpressionVisitor class to reduce duplication between SnQL and Clickhouse Visitors.

ayirr7 committed:
Modified SnQL visitor, reduced duplication.

ayirr7 committed:
Cosmetic changes to Clickhouse Visitor.

ayirr7 committed:
Merge branch 'master' into feat/snqlvisitor

fpacifici commented on snuba/query/snql/__init__.py:
```diff
@@ -1,23 +1,187 @@
+        return Query({}, None, collect, None,)
+
+    def visit_function_name(self, node: Node, visited_children: Iterable[Any]) -> str:
+        return ExpressionVisitor.visit_function_name(self, node, visited_children)
```
This is not supposed to work. You should not pass self into an instance method of a class.  Python takes care of that. 
If you want to call a method without having an instance of the class you have two alternatives.
1) Just don't create the class. You do not need to define the ExpressionVisitor class, all the methods are class methods (do not rely on the instance), so why not just defining them as functions of the expression_visitor module, and here you import directly the functions.
2) define them as class methods of the class ExpressionVisitor. Which means you decorate the method with `@classmethod` and, from here you do not pass self when calling the method. 

I would suggest option 1 since you do not have any method in ExpressionVisitor that depends on the state of the class, so there is no reason for having the class.

ayirr7 replied:
I applied option 1. It seems to me that the issue I was facing earlier still exists: for some reason, there are 2 identical (correct) Query objects being produced.

fpacifici commented on snuba/query/snql/expression_visitor.py:
```diff
@@ -0,0 +1,196 @@
+
+
+class ExpressionVisitor:
+    def visit_function_name(left, node: Node, visited_children: Iterable[Any]) -> str:
```
Shouldn't these methods be used by the old expression parser as well ? https://github.com/getsentry/snuba/blob/master/snuba/query/parser/expressions.py#L117

Or do you plan to make that change later ?

ayirr7 replied:
Yes, I think I can fix the init file first and then apply it to Clickhouse Visitor.

fpacifici reviewed (changes requested):
The direction is good, but please see my comments around the ExpressionVisitor inline.

ayirr7 committed:
Removed ExpressionVisitor class.

ayirr7 committed:
Removed Visitor class.

ayirr7 committed:
Renamed visitor functions.

fpacifici commented on snuba/query/snql/expression_visitor.py:
```diff
@@ -0,0 +1,205 @@
+    return term
+
+
+def exp_visit_function_name(left, node: Node, visited_children: Iterable[Any]) -> str:
```
I don't think you need to prefix all function names with `exp_`.

fpacifici commented on snuba/query/snql/expression_visitor.py:
```diff
@@ -0,0 +1,205 @@
+    return term
+
+
+def exp_visit_low_pri_arithmetic(
+    left, node: Node, visited_children: Tuple[Any, Expression, Any, LowPriArithmetic],
+) -> Expression:
+    _, term, _, exp = visited_children
+
+    return get_arithmetic_expression(term, exp)
+
+
+def exp_visit_high_pri_arithmetic(
+    left, node: Node, visited_children: Tuple[Any, Expression, Any, HighPriArithmetic],
+) -> Expression:
+    _, term, _, exp = visited_children
+
+    return get_arithmetic_expression(term, exp)
```
I don't think you need both these methods. they do the same. You can make the visitor call directly `get_arithmetic_expression`.

fpacifici reviewed (changes requested):
A few comments inline, also please add the unit tests to ensure the parser works.

ayirr7 committed:
Modified init to account for unit tests.

ayirr7 committed:
Removed exp_ prefix from visitor file.

ayirr7 committed:
Added Unit tests for SnQL Visitor.

ayirr7 committed:
Removed duplication by using SnQL expression visitor.

ayirr7 commented on snuba/query/snql/__init__.py:
```diff
@@ -1,23 +1,205 @@
+
+snql_grammar = Grammar(
+    fr"""
+    query_exp             = match_clause? where_clause? collect_clause? having_clause? order_by_clause?
```
So as of now I am keeping them optional just to test the COLLECT clause individually. As I add support for all the clauses, I will remove the optional thing.

fpacifici replied:
ok, please keep the match still because we are not adding the content of that any time soon.

ayirr7 commented:
Updates:

- Added unit testing for COLLECT clause
- Removed duplication for both Clickhouse Visitor and SnQL Visitor

Next:

- Expand grammar to handle other clauses
- Add corresponding unit tests

fpacifici commented on tests/query/snql/test_query.py:
```diff
@@ -0,0 +1,66 @@
+
+test_cases = [
+    pytest.param(
+        "COLLECT 4-5, 3*g(c), c",
```
Shouldn't the grammar reject a query that does not have a MATCH clause ?
Even if you do not parse the content of the match, the query should still be well formatted.

fpacifici commented on snuba/query/snql/__init__.py:
```diff
@@ -1,28 +1,194 @@
+    Builds Snuba AST expressions from the Parsimonious parse tree.
+    """
+
+    def visit_query_exp(self, node, visited_children):
```
Please add the missing type signatures to all methods.

ayirr7 replied:
I will get to this asap! Just wanted to update the GROUP BY and ORDER BY additions first.

fpacifici replied:
still types missing.

fpacifici commented on snuba/query/snql/__init__.py:
```diff
@@ -1,28 +1,194 @@
+        ret: List[SelectedExpression] = []
+        if not isinstance(column_list, Node):
+            if not isinstance(column_list, (list, tuple)):
+                ret.append(SelectedExpression(None, column_list))
+            else:
+                for p in column_list:
+                    ret.append(SelectedExpression(None, p))
```
This is missing one part: you are not assigning the names to the SerlectedExpression objects. SelectedExpression.name should be the name the user expects in the output. It works like for SQL:
```
SELECT a, b, f(x), g(k) AS alias....
```
Would look like in the output
```
a | b | f(x) | alias
------------------
values. ....
```

So the algorithm should work as:
1) if the parsed expression (column_list at line 152 and p here) has an alias then the name is the alias.
2) if not, you should use the string the user provided, which you need to preserve the raw expression, which should be the raw string for low_pri_aritthmetic from the `collect_columns       = low_pri_arithmetic space* comma space*` rule. The raw string should be accessible from the parsed node.

fpacifici commented on snuba/query/snql/__init__.py:
```diff
@@ -1,28 +1,194 @@
-    return Query({}, None,)
+    exp_tree = snql_grammar.parse(body)
+    parsed_exp = SnQLVisitor().visit(exp_tree)
+    selected_columns = parsed_exp.get_selected_columns_from_ast()
+    return Query({}, None, selected_columns,)
```
Why can't you just return parsed_exp ?

fpacifici commented on snuba/query/snql/expression_visitor.py:
```diff
@@ -0,0 +1,197 @@
+
+def generic_visit(node: Node, visited_children: Any) -> Any:
+    if isinstance(visited_children, list) and len(visited_children) == 1:
+        return visited_children[0]
```
Could you please add the comment back? It is still valid.

fpacifici reviewed (changes requested):
I think there are still a couple of areas to fix.

ayirr7 committed:
Add GROUPBY clause.

ayirr7 committed:
Restored an old comment.

ayirr7 committed:
Add GROUPBY clause to unit test.

ayirr7 committed:
Merge branch 'master' into feat/snqlvisitor

ayirr7 committed:
Add ORDERBY clause.

ayirr7 committed:
Add ORDERBY unit tests.

ayirr7 commented:
Updates: 
- Assigned names to SelectedExpressions
- Added support for (GROUP) BY, ORDER BY in grammar and visitor
- Added unit test for above 

Next Steps:
- Conditions (individual first, then combinations)

fpacifici commented on snuba/query/snql/__init__.py:
```diff
@@ -1,28 +1,255 @@
+        if not isinstance(column_list, Node):
+            if not isinstance(column_list, (list, tuple)):
+                if column_list.alias is None:
+                    ret.append(SelectedExpression(node.text, column_list))
```
I am not sure node.text is the right way to extract the raw string here. It seems to me that it would not work if the collect clause contains only two expressions. We would ennd here but node.text would contain the whole raw collect expression. 

The safe way of doing this is by returning the raw string together with the Expression object from `visit_collect_clause`. Then you just need to pick the raw string provided there.

fpacifici commented on snuba/query/snql/__init__.py:
```diff
@@ -1,28 +1,255 @@
-
-    return Query({}, None,)
+    exp_tree = snql_grammar.parse(body)
+    parsed_exp = SnQLVisitor().visit(exp_tree)
+    return parsed_exp
```
return SnQLVisitor().visit(exp_tree)

fpacifici commented on snuba/query/snql/__init__.py:
```diff
@@ -1,28 +1,255 @@
+        elif order.text == "DESC":
+            ret.append(OrderBy(OrderByDirection.DESC, right_order))
+
+        return ret
```
Why would you ever get to this statement ?

ayirr7 replied:
Grammar has the structure of:

order_list            = order_columns* low_pri_arithmetic ("ASC"/"DESC")

so the righthand side (low_pri_arithmetic ("ASC"/"DESC") is not being processed as order_columns. In order to add it to the order_by list, then, we need to process it in visit_order_list and take a look at whether it is ASC/DESC.

ayirr7 committed:
Added alias logic.

ayirr7 committed:
Merge branch 'master' into feat/snqlvisitor

fpacifici commented on snuba/query/snql/__init__.py:
```diff
@@ -1,28 +1,259 @@
+    def visit_collect_list(self, node, visited_children):
+        column_list, right_column = visited_children
+        ret: List[SelectedExpression] = []
+        if not isinstance(column_list, Node):
```
please add a comment explaining this condition. Who has no context does not know why column_list would ever be a Node

ayirr7 replied:
Added.

ayirr7 committed:
Modified approach for alias names, added checks for missing cases in Visitor, type hints.

ayirr7 committed:
Test case.

ayirr7 committed:
Merge branch 'master' into feat/snqlvisitor

fpacifici commented on snuba/query/snql/__init__.py:
```diff
@@ -1,28 +1,296 @@
+    ) -> OrderBy:
+        column, order, _, _, _ = visited_children
+
+        if order.text == "ASC":
+            order_object = OrderBy(OrderByDirection.ASC, column)
+        elif order.text == "DESC":
+            order_object = OrderBy(OrderByDirection.DESC, column)
+
+        return order_object
```
```
direction = OrderByDirection.ASC if order.text == "ASC" else OrderByDirection.DESC
return OrderBy(direction, column)
```

fpacifici commented on snuba/query/snql/__init__.py:
```diff
@@ -1,28 +1,296 @@
+        self, node: Node, visited_children: Tuple[Expression, Any]
+    ) -> SelectedExpression:
+        exp, _ = visited_children
+        return SelectedExpression(node.text.strip(), exp)
```
Do you actually need strip() here ?

ayirr7 replied:
Yes. otherwise there are trailing spaces. I think the reason for these spaces is the 'comma, space' format.

fpacifici reviewed (approved):
excellent. thanks.
